### ISSUE TITLE
git-rebase-mode font-lock improvements

### ISSUE THREAD

**Issue by fice-t** [magit/magit#2769]
- [x] I think the face `git-rebase-hash` should inherit from `magit-hash` in case a theme only sets the latter.

- [ ] It would be nice if the commits in the comment (i.e. `# Rebase d436c18..ccc47eb onto d436c18`) were highlighted like the branches are in the commit buffer.

- [ ] Perhaps the `Commands:` could be highlighted like `Changes to be committed:` is in the commit buffer?

- [ ] While posting this I thought it might be nice if the command keybindings in the comment were highlighted like in `magit-popup`. Do you think that would be gaudy?

- [ ] This is unrelated to font-lock, but I also think that `p` and `n` should appear in the commands list aside `C-p` and `C-n` respectively.

`Magit 20160902.1451, Git 2.9.3, Emacs 25.1.3, gnu/linux`

**Comment by fice-t** [comment]
This is unrelated to font-lock, but I also think that `p` and `n` should appear in the commands list aside `C-p` and `C-n` respectively.

**Comment by tarsius** [comment]
> hash

Done that

> p and n

Unfortunately that's not possible using the current implementation.

> font-lock

I am not sure about the rest. Personally I have set `git-rebase-show-instructions` to `nil`. I suspect that many users haven't done that but that going full fruit-salad here would push them over the edge.

**Comment by tarsius** [comment]
> While posting this I thought it might be nice if the command keybindings in the comment were highlighted like in `magit-popup`.

Coming with a font-lock keyword regexp that avoided false-positives would be hard.

**Comment by fice-t** [comment]
@tarsius Thanks.

> Coming with a font-lock keyword regexp that avoided false-positives would be hard.

Would you accept a PR that did this (I might have a working regexp)? I could make it optional if it's too much like fruit salad.

Also, if I may ask 2 questions here:

Is there a reason why `d` also doesn't drop the commit at point?

Why do both keybindings for the reword/edit commands show up, but only one for the drop command? Does the Control modifier limit the ability to show the other bindings?

**Comment by tarsius** [comment]
> Would you accept a PR that did this (I might have a working regexp)?

Yes, but you would have to go further than the regexp.

> Why do both keybindings for the reword/edit commands show up, but only one for the drop command? Does the Control modifier limit the ability to show the other bindings?

In one case `substitute-command-keys` is used and in the other `(mapconcat #'key-description ...)`. This happens in `git-rebase-mode-show-keybindings` where you would also have to make the other changes beside coming up with a regexp.

Things to consider:

* Maybe just discard the binding information printed altogether and print all bindings the same way.
  * I am unsure whether I prefer printing only the "best" binding or all.
    * Whether we succeed in determining the best binding would have a huge influence on that.
* Probably indent all bindings with one or two extra spaces.
  * Looks nicer.
  * Easier to apply font-lock keywords.
    * Even if we don't discard git's output we are in a better position to avoid false-positives in this function than during font-lock because we can use the context  (after "# Commands:", before "#".
    * The font-lock regexp then becomes simpler, the extra space(s) should help a lot avoiding false-positives.
  * We used to indent but there was some issue with some git version indenting and others not.
    * I went with "removing whitespace if it is there", but going with "insert whitespace if it is not there yet" would also have worked.
    * And maybe that isn't relevant anymore because the git versions that do it differently from the current version is older than 1.9.4, which now is the minimal version.
    * You would have to dig through history to find out.
    * Or maybe if we just discarded git's binding output, then this wouldn't be relevant anymore. But that could come with other complications.

Basically, I have thought about these things before, but always found something better to do ;-) But if you want to do it, then please go ahead.

**Comment by tarsius** [comment]
> Is there a reason why `d` also doesn't drop the commit at point?

For consistency with Magit where `d` is "diff". Also `k` is "kill" a lot in Emacs, so that's consistent with that too. (It does however cause issues for me, since, as I mentioned, I use `k` (or `C-k`) to go to the next line. So I use `,` or `C-,` to kill instead (consistently across all of Emacs). It's un uphill battle. ;-)
